### PR TITLE
fsck.ocfs2: do not try locking after replaying journals if -F is given

### PR DESCRIPTION
**Pull request by kastixx**
With -F option, `fsck.ocfs2` skips locking down the cluster, but not if
the journals are replayed. If it has to replay the journals then it
tries to lock down the cluster regardless of -F.

This fix adds an extra check that prevents locking down the cluster
after journal replay if -F is given.

### CODE REVIEW

File: fsck.ocfs2/fsck.c [hunk 1]
```diff
@@ -551,7 +551,9 @@ static errcode_t maybe_replay_journals(o2fsck_state *ost, char *filename,
 		goto out;
 	}
 
-	ret = fsck_lock_fs(ost);
+	if (!ost->ost_skip_o2cb) {
```
Comment (YuraKotov) [inline fsck.ocfs2/fsck.c:554]: I think it should be:
``` c
	if (open_flags & OCFS2_FLAG_RW && !ost->ost_skip_o2cb &&
	    !ocfs2_mount_local(ost->ost_fs)) {
```

Comment (kastixx) [reply fsck.ocfs2/fsck.c]: Yes, thanks, added missing checks

**Review by zhaohem — APPROVED** [review]
@biger410 @josephhz  this patch fixes the problem which is introduced by commit 736f2108c24c93 ("ocfs2: fix cluster unlocked if journal is replayed").
the fix follows the exist code logic in main(), IMO it's correct.
Please accept this patch.

**Comment by josephhz** [comment]
I commented before, this patch should be updated and then resend.

**Comment by zhaohem** [comment]
Hello kastixx & YuraKotov,

I am waiting for this patch to be accepted. could you have time to resend v2 patch?

**Comment by kastixx** [comment]
> I commented before, this patch should be updated and then resend.

@josephhz sorry, did not receive your comment. Could you please RE-post it?

**Comment by josephhz** [comment]
image

**Comment by zhaohem** [comment]
This patch makes sense to real world. Recently one of SUSE customers reported a bug.
They cloned a disk data to another disk, two disks on the same machine and the source disk was mounted & working. Some reasons made the cloned disk include dirty journal.  When they ran "fsck.ocfs2 -F" which caused lock failed.

**Comment by kastixx** [comment]
> image

@josephhz Thanks!
I've included your suggestion (removed redundant `OCFS2_FLAG_RW` condition) and squashed three commits into one -- is this OK?

**Comment by kastixx** [comment]
BTW, I didn't see your comment because it is still unpublished (it is marked as Pending on the screenshot). You need to click the green "Review changes" button in the top right corner on "Files changed" tab and click "Submit review" in the pop-up window in order to publish all pending comments. This is a common mistake with GitHub, I've made it many times myself. The UI is not perfectly intuitive here.

**Comment by josephhz** [comment]
> BTW, I didn't see your comment because it is still unpublished (it is marked as Pending on the screenshot). You need to click the green "Review changes" button in the top right corner on "Files changed" tab and click "Submit review" in the pop-up window in order to publish all pending comments. This is a common mistake with GitHub, I've made it many times myself. The UI is not perfectly intuitive here.

Sorry, my mistake... Don't often review on github...
Since we can call fsck_unlock_fs() directly, I think do it the same way and the code looks more consistent.
How about we change it in the following way? 

diff --git a/fsck.ocfs2/fsck.c b/fsck.ocfs2/fsck.c
index 757224f4..0c430a5b 100644
--- a/fsck.ocfs2/fsck.c
+++ b/fsck.ocfs2/fsck.c
@@ -707,6 +707,10 @@ static errcode_t fsck_lock_fs(o2fsck_state *ost)
 {
 	errcode_t ret;
 
+	if ((ost->ost_fs->fs_flags & OCFS2_FLAG_RW) && !ost->ost_skip_o2cb &&
+	    !ocfs2_mount_local(ost->ost_fs))
+		return 0;
+
 	ret = o2cb_init();
 	if (ret) {
 		com_err(whoami, ret, "while initializing the cluster");
@@ -986,13 +990,10 @@ int main(int argc, char **argv)
 		goto out;
 	}
 
-	if (open_flags & OCFS2_FLAG_RW && !ost->ost_skip_o2cb &&
-	    !ocfs2_mount_local(ost->ost_fs)) {
-		ret = fsck_lock_fs(ost);
-		if (ret) {
-			fsck_mask |= FSCK_ERROR;
-			goto close;
-		}
+	ret = fsck_lock_fs(ost);
+	if (ret) {
+		fsck_mask |= FSCK_ERROR;
+		goto close;
 	}
 
 	printf("Checking OCFS2 filesystem in %s:\n", filename);

BTW, you missed your signed-of-by.

**Comment by kastixx** [comment]
>  Since we can call fsck_unlock_fs() directly, I think do it the same way and the code looks more consistent.

Maybe do it this way then?
```patch
--- a/fsck.ocfs2/fsck.c
+++ b/fsck.ocfs2/fsck.c
@@ -551,9 +551,7 @@ static errcode_t maybe_replay_journals(o2fsck_state *ost, char *filename,
 		goto out;
 	}
 
-	if (!ost->ost_skip_o2cb && !ocfs2_mount_local(ost->ost_fs)) {
-		ret = fsck_lock_fs(ost);
-	}
+	ret = fsck_lock_fs(ost);
 out:
 	return ret;
 }
@@ -709,6 +707,10 @@ static errcode_t fsck_lock_fs(o2fsck_state *ost)
 {
 	errcode_t ret;
 
+	if (!(ost->ost_fs->fs_flags & OCFS2_FLAG_RW) || !ost->ost_skip_o2cb ||
+	    ocfs2_mount_local(ost->ost_fs))
+		return 0;
+
 	ret = o2cb_init();
 	if (ret) {
 		com_err(whoami, ret, "while initializing the cluster");
@@ -988,13 +990,10 @@ int main(int argc, char **argv)
 		goto out;
 	}
 
-	if (open_flags & OCFS2_FLAG_RW && !ost->ost_skip_o2cb &&
-	    !ocfs2_mount_local(ost->ost_fs)) {
-		ret = fsck_lock_fs(ost);
-		if (ret) {
-			fsck_mask |= FSCK_ERROR;
-			goto close;
-		}
+	ret = fsck_lock_fs(ost);
+	if (ret) {
+		fsck_mask |= FSCK_ERROR;
+		goto close;
 	}
 
 	printf("Checking OCFS2 filesystem in %s:\n", filename);
```

**Comment by josephhz** [comment]
Oh, when move it into fsck_lock_fs(), the check condition should be:
if (!(ost->ost_fs->fs_flags & OCFS2_FLAG_RW) || ost->ost_skip_o2cb ||
    ocfs2_mount_local(ost->ost_fs))

**Comment by kastixx** [comment]
> BTW, you missed your signed-of-by.

I will add sign-off to the final version of the patch. This 

> Oh, when move it into fsck_lock_fs(), the check condition should be: if (!(ost->ost_fs->fs_flags & OCFS2_FLAG_RW) || ost->ost_skip_o2cb || ocfs2_mount_local(ost->ost_fs))

yep, noticed that too, thanks

File: fsck.ocfs2/fsck.c [hunk 2]
```diff
@@ -707,6 +707,10 @@ static errcode_t fsck_lock_fs(o2fsck_state *ost)
 {
 	errcode_t ret;
 
+	if (!(ost->ost_fs->fs_flags & OCFS2_FLAG_RW) || !ost->ost_skip_o2cb ||
```
Comment (josephhz) [inline fsck.ocfs2/fsck.c:710]: It should be ost->ost_skip_o2cb.

Comment (kastixx) [reply fsck.ocfs2/fsck.c]: fixed, thanks

**Comment by josephhz** [comment]
Looks good now.

**Pull request merged**